product-service/src/libs/data-service.ts

```diff
@@ -1,10 +1,118 @@
-import { productsData } from '@mocks/products';
-import { IProduct, IProductsData } from 'src/types/api-types';
+import { Product, ProductsData, Stock } from 'src/types/api-types';
+import {
+  AttributeValue,
+  DynamoDBClient,
+  PutItemCommand,
+  PutItemCommandInput,
+  QueryCommand,
+  QueryInput,
+  ScanCommand,
+  ScanCommandInput
+} from '@aws-sdk/client-dynamodb';
 
-export const getProducts = async (): Promise<IProductsData> => {
-  return Promise.resolve(productsData);
+const dynamodb = new DynamoDBClient({ region: process.env.REGION || 'us-east-1' });
+
+function deserializeDynamoDBItem(item: Record<string, AttributeValue>) {
+  const deserializedItem = {};
+  for (const [attributeName, value] of Object.entries(item)) {
+    const attributeType = Object.keys(value)[0];
+
+    if (attributeType === 'S') {
+      deserializedItem[attributeName] = value.S;
+    } else if (attributeType === 'N') {
+      deserializedItem[attributeName] = parseFloat(value.N);
+    }
+  }
+
+  return deserializedItem;
+}
+
+async function queryProducts(): Promise<Product[]> {
+  const params: ScanCommandInput = {
+    TableName: process.env.PRODUCTS_TABLE_NAME || 'products'
+  };
+
+  const data = await dynamodb.send(new ScanCommand(params));
+
+  return data.Items.map((item) => deserializeDynamoDBItem(item)) as Product[];
+}
+
+async function queryStock(productId: string): Promise<Stock> {
+  const params: QueryInput = {
+    TableName: process.env.STOCKS_TABLE_NAME || 'stocks',
+    KeyConditionExpression: 'product_id = :id',
+    ExpressionAttributeValues: {
+      ':id': { S: productId }
+    }
+  };
+
+  const data = await dynamodb.send(new QueryCommand(params));
+
+  return deserializeDynamoDBItem(data.Items[0]) as Stock;
+}
+
+export const getProducts = async (): Promise<ProductsData> => {
+  try {
+    const products = await queryProducts();
+
+    const productsWithStock = await Promise.all(
+      products.map(async (product) => {
+        const stock = await queryStock(product.id);
+
+        return { ...product, count: (stock && stock.count) || 0 };
+      })
+    );
+
+    return { products: productsWithStock };
+  } catch (error) {
+    console.error(error);
+    throw error;
+  }
+};
+
+export const getProductById = async (productId): Promise<Product | undefined> => {
+  const { products } = await getProducts();
```

Comment:
**Concern**: We don't need all products, we need only one. I guess you should implement something like `queryStock` but for productTable.

Reply:
you're right I probably need to have a separate query for this product, didn't take that into account